extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -66,23 +73,56 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
                     public Uni<Boolean> apply(Permission permission) {
                         return securityIdentity.checkPermission(permission);
                     }
-                })
-                .build();
+                }).build();
     }
 
     public Uni<SecurityIdentity> authenticate(RoutingContext context,
             IdentityProviderManager identityProviderManager,
             DefaultTenantConfigResolver resolver) {
         Cookie sessionCookie = context.request().getCookie(SESSION_COOKIE_NAME);
+        TenantConfigContext configContext = resolver.resolve(context, true);
 
         // if session already established, try to re-authenticate
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
             return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
                     .map(new Function<SecurityIdentity, SecurityIdentity>() {
                         @Override
-                        public SecurityIdentity apply(SecurityIdentity securityIdentity) {
-                            return augmentIdentity(securityIdentity, tokens[1], tokens[2], context);
+                        public SecurityIdentity apply(SecurityIdentity identity) {
+                            if (isLogout(context, configContext)) {
+                                throw redirectToLogoutEndpoint(context, configContext, idToken);
+                            }
+
+                            return augmentIdentity(identity, accessToken, refreshToken, context);
+                        }
+                    }).on().failure().recoverWithItem(new Function<Throwable, SecurityIdentity>() {
+                        @Override
+                        public SecurityIdentity apply(Throwable throwable) {
+                            if (throwable instanceof AuthenticationRedirectException) {
+                                throw AuthenticationRedirectException.class.cast(throwable);
+                            }
+
+                            Throwable cause = throwable.getCause();
+
+                            // we should have proper exception hierarchy to represent token expiration errors
+                            if (cause != null && !cause.getMessage().equalsIgnoreCase("expired token")) {
+                                throw new AuthenticationFailedException(throwable);
```

Comment:
Pretty sure now this should be `AuthenticationCompletionException` to get 401 out if it is something anything but expiry related.

Reply:
I think I see what you want. Avoid redirects in case of "unknown" failures. Will change.